File: packages/fundamentals/hooks/native/kotlin/com/pyreon/runtime/PyreonSpeech.kt
// PyreonSpeech — the TextToSpeech side of `@pyreon/hooks`' useSpeech.
// Mirror of PyreonSpeech.swift; see that header for why rate/pitch/voice are
// out of scope.

package com.pyreon.runtime

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf

/** The platform half. Swapped for a fake in tests. */
public interface SpeechSynth {
    public val isAvailable: Boolean
    public fun speak(text: String)
    public fun cancel()
}

public class PyreonSpeech(private val synth: SpeechSynth) {
    public val speaking: MutableState<Boolean> = mutableStateOf(false)

    public val supported: Boolean get() = synth.isAvailable

    public fun speak(text: String): Boolean {
        if (!synth.isAvailable || text.isEmpty()) return false
        // Cancel first — queueing is the platform default.
        synth.cancel()
        synth.speak(text)
        speaking.value = true
        return true
    }

    public fun stop() {
        synth.cancel()
        speaking.value = false
    }
}

File: packages/fundamentals/hooks/native/swift/PyreonSpeech.swift
// PyreonSpeech — the AVSpeechSynthesizer side of `@pyreon/hooks`' useSpeech.
//
// Rate, pitch and voice selection are out of scope: the three platforms
// disagree on their ranges and on how voices are identified, so one name
// would mean three different things. Plain speech is what crosses.
//
// The synthesiser is injected so the state machine is testable with no
// AVFoundation and no audio device.

import Foundation
import Observation

/// The platform half. Swapped for a fake in tests.
public protocol SpeechSynth: AnyObject {
    var isAvailable: Bool { get }
    func speak(_ text: String)
    func cancel()
}

@available(iOS 17.0, macOS 14.0, *)
@Observable
public final class PyreonSpeech {
    public private(set) var speaking: Bool = false
    private let synth: SpeechSynth

    public var supported: Bool { synth.isAvailable }

    public init(synth: SpeechSynth) {
        self.synth = synth
    }

    @discardableResult
    public func speak(_ text: String) -> Bool {
        guard synth.isAvailable, !text.isEmpty else { return false }
        // Cancel first — queueing is the platform default, so a second press
        // would otherwise talk over itself rather than replacing.
        synth.cancel()
        synth.speak(text)
        speaking = true
        return true
    }

    public func stop() {
        synth.cancel()
        speaking = false
    }

    deinit {
        // Speech outliving its view talks over the next screen.
        synth.cancel()
    }
}
